app/graphql/filters/available_rocks.rb

```diff
@@ -0,0 +1,43 @@
+require 'search_object'
+require 'search_object/plugin/graphql'
+  class Filters::AvailableRocks
+    # include SearchObject for GraphQL
+    include SearchObject.module(:graphql)
+
+    # scope is starting point for search
+    scope { User.all }
+
+    type types[Types::UserType]
+
+    # inline input type definition for the advanced filter
+    class UserFilter < ::Types::BaseInputObject
+      argument :OR, [self], required: false
+      argument :program, String, required: false
+      argument :module, Integer, required: false
+    end
+
+    # when "filter" is passed "apply_filter" would be called to narrow the scope
+    option :filter, type: UserFilter, with: :apply_filter
+
+    # apply_filter recursively loops through "OR" branches
+    def apply_filter(scope, value)
+      branches = normalize_filters(value).reduce { |a, b| a.or(b) }
+      scope.merge branches
+    end
+
+    def normalize_filters(value, branches = [])
+      scope = User.all
+      scope = scope.where('program LIKE ?', "%#{value[:program]}%") if value[:program]
+      scope = scope.where('mod = ?', value[:module]) if value[:module]
+      scope = scope.where(rock_opt_in: true)
+      users = scope.left_outer_joins(:pebble_roles).where(rock_and_pebbles: { rock_id: nil } )
+      rock_and_pebbles = scope.left_outer_joins(:pebble_roles).where(rock_and_pebbles: {active: true}).group(:id).having('count(pebble_id) < 2')
+
+      scope = users + rock_and_pebbles
+      branches << scope
+
+      value[:OR].reduce(branches) { |s, v| normalize_filters(v, s) } if value[:OR].present?
+
+      branches
+    end
+  end
```

Comment:
I think I'm going to need a walk-through of this whole file, my head is spinning.

Reply:
Here is a link to an example of this gem being used: https://www.howtographql.com/graphql-ruby/7-filtering/
I anytime a query has the keyword 'filter' then I believe it needs to follow a similar template and use the search_object gem. The original paired app had a query with the key word filter which is how I came across this. I'm not entirely sure it is necessary though. I think it might be possible to have the query as a mutation and just do the active record query in the resolver.

Reply:
My only thought with that is mutations are probably only supposed to be used to create or edit something and filter queries aren't doing that. The only lines you really need to be aware of are 29 - 36. That is where the 'filtering' is being accomplished.

Reply:
That tutorial and your comments helped me make sense of it, thanks! 

Seems like filters are helpful when there are multiple optional search parameters or you want to give open-ended access to search a resource (like through a search bar); whereas in a query, the search and the arguments are going to be more rigid. 

In this case, unless we plan on making either "Mod" or "Program" an optional search restriction, I think the filter functionality is more than we need and we could scale this back to be a simple query with mod and program as required arguments. If it's working, though, I don't think there's an immediate need to change it; the only impact it would have is that the FE will need to know the alternate "filter" syntax. What are your thoughts on that?

Reply:
I agree that it is probably overkill in terms of functionality but the gem isn't hard to work with once you understand the syntax and since the front end already has a filter query that it is working with there should be a good example to follow. I think it's fine to leave it as is. :)